src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs

```diff
@@ -333,5 +379,125 @@ private SourceRepository GetSourceRepository(PackageSource source)
                 Owners = nuspec.GetOwners()
             };
         }
+
+        public async Task<bool> DownloadPackageToDirectoryAsync(string path, NuGetReference nugetReference, string source, CancellationToken token)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                throw new ArgumentException($"'{nameof(path)}' cannot be null or empty.", nameof(path));
+            }
+
+            if (nugetReference is null)
+            {
+                throw new ArgumentNullException(nameof(nugetReference));
+            }
+
+            using var stream = GetCachedPackage(nugetReference) ?? await GetPackageStream(nugetReference, source, token).ConfigureAwait(false);
+
+            if (stream is null)
+            {
+                _logger.LogWarning("Could not find {Package} [{Version}]", nugetReference.Name, nugetReference.Version);
+                return false;
+            }
+
+            try
+            {
+                using var zip = new ZipArchive(stream);
+
+                zip.ExtractToDirectory(path);
+
+                return true;
+            }
+            catch (InvalidDataException)
+            {
+            }
+            catch (IOException)
+            {
+            }
```

Comment:
Why empty catches? need to log something here.

Reply:
There's a generalized error below these. If successful, nothing beyond the try will be hit; on the caught exceptions it does. That's so there's less duplication as we just want to let them know it's amalformed package.